packages/git/src/browser/dirty-diff/dirty-diff-manager.ts

```diff
@@ -0,0 +1,251 @@
+                model.enabled = enabled;
+                model.update();
+            }
+        }
+    }
+
+    protected async allModels(): Promise<DirtyDiffModel[]> {
+        const models = [];
+        const uris = this.models.keys();
+        for (const uri of uris) {
+            const reference = await this.models.acquire(uri);
+            models.push(reference.object);
+            reference.dispose();
+        }
+        return models;
+    }
+
+    protected async getModel(uri: string): Promise<DirtyDiffModel | undefined> {
+        if (this.models.has(uri)) {
+            const reference = await this.models.acquire(uri);
+            const model = reference.object;
+            reference.dispose();
+            return model;
+        }
+        return undefined;
+    }
+
+}
+
+export interface DirtyDiffUpdate extends DirtyDiff {
+    readonly uri: string;
+}
+
+export class DirtyDiffModel implements Disposable {
+
+    enabled = true;
+
+    protected dirty = true;
+    protected staged: boolean;
+    protected previousContent: string[];
+    protected currentContent: string[];
+
+    protected readonly onDirtyDiffUpdateEmitter = new Emitter<DirtyDiffUpdate>();
+    readonly onDirtyDiffUpdate: Event<DirtyDiffUpdate> = this.onDirtyDiffUpdateEmitter.event;
+    protected readonly updateDelayer = new DirtyDiffModel.Throttler(200);
+
+    constructor(
+        readonly uri: string,
+        protected readonly readGitResource: DirtyDiffModel.GitResourceReader
+    ) { }
+
+    update(): void {
+        const enabled = this.enabled && this.dirty;
+        const previous = enabled ? this.previousContent : [];
+        const current = enabled ? this.currentContent : [];
+        if (!previous || !current) {
+            return;
+        }
+        const uri = this.uri;
+        this.updateDelayer.push(() => {
+            const dirtyDiff = enabled
+                ? DirtyDiffModel.computeDirtyDiff(previous, current)
+                : <DirtyDiff>{ added: [], removed: [], modified: [] };
+            const dirtyDiffUpdate = <DirtyDiffUpdate>{ uri, ...dirtyDiff };
+            this.onDirtyDiffUpdateEmitter.fire(dirtyDiffUpdate);
+        });
+    }
+
+    handleDocumentChanged(documentContent: string): void {
+        this.currentContent = DirtyDiffModel.splitLines(documentContent);
+        if (!this.previousContent) {
+            this.previousContent = [...this.currentContent];
+        }
+        this.update();
+    }
+
+    async handleGitStatusUpdate(changes: GitFileChange[]): Promise<void> {
+        const modified = changes.filter(c => c.status === GitFileStatus.Modified)[0];
+        const newStaged = changes.filter(c => c.status === GitFileStatus.New && c.staged)[0];
+        const newUnstaged = changes.filter(c => c.status === GitFileStatus.New && !c.staged)[0];
+        if (modified || newStaged) {
+            this.dirty = true;
+            this.staged = !!newStaged || (modified && (modified.staged || false));
```

Comment:
Hopefully, someone will understand this logic a week from now... because I cannot.
Why do not you use `changes.some`? Probably it's just me, but I really cannot see what it supposed to do.

Reply:
Thanks @kittaakos! I see the point. Could you please re-read the updated version. I've improved the naming.